tsdb/db.go

```diff
@@ -638,6 +638,21 @@ func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs
 		return nil, err
 	}
 
+	defer func() {
+		// Close files if startup fails somewhere
+		if err != nil {
+			if er := db.head.Close(); er != nil {
```

Comment:
you should check for nil head.

Reply:
We have this right above the deferred function
```
db.head, err = NewHead(r, l, wlog, rngs[0], dir, db.chunkPool, opts.StripeSize, opts.SeriesLifecycleCallback)
if err != nil {
	return nil, err
}
```

Can `NewHead` return a nil head? I assumed that this behavior would only occur with err != nil, in that case, we don't need to check for a nil head

But again, I'm a beginner here, please correct me if I'm wrong

Reply:
you are right.

I think we need to change NewHead too, to close the head in a defer if it is returning an error.

Reply:
I've just taken a look at the NewHead function and I think there is no need.

The only resources opened are [chunk disks](https://github.com/prometheus/prometheus/blob/a7429890475045b1304cd690464b859af1ae1ce0/tsdb/head.go#L333-L336), which are closed by [db.head.Close()](https://github.com/prometheus/prometheus/blob/a7429890475045b1304cd690464b859af1ae1ce0/tsdb/head.go#L1441).

Reply:
If NewHead errors you do not call that

Reply:
don't we need to close the wal?

Reply:
you're right with both comments! I've misplaced the deferred function

should be right now